src/controllers/hid/hidio.cpp

```diff
@@ -0,0 +1,259 @@
+#include "controllers/hid/hidio.h"
+
+#include <hidapi.h>
+
+#include "controllers/defs_controllers.h"
+#include "controllers/hid/legacyhidcontrollermappingfilehandler.h"
+#include "moc_hidio.cpp"
+#include "util/string.h"
+#include "util/time.h"
+#include "util/trace.h"
+
+namespace {
+constexpr int kReportIdSize = 1;
+constexpr int kMaxHidErrorMessageSize = 512;
+} // namespace
+
+HidIoReport::HidIoReport(const unsigned char& reportId,
+        hid_device* device,
+        const mixxx::hid::DeviceInfo&& deviceInfo,
+        const RuntimeLoggingCategory& logOutput)
+        : m_reportId(reportId),
+          m_pHidDevice(device),
+          m_deviceInfo(std::move(deviceInfo)),
+          m_logOutput(logOutput),
+          m_lastSentOutputreport("") {
+}
+
+HidIoReport::~HidIoReport() {
+}
+
+void HidIoReport::sendOutputReport(QByteArray data) {
+    auto startOfHidWrite = mixxx::Time::elapsed();
+    if (!m_lastSentOutputreport.compare(data)) {
+        qCDebug(m_logOutput) << "t:" << startOfHidWrite.formatMillisWithUnit()
+                             << " Skipped identical Output Report for" << m_deviceInfo.formatName()
+                             << "serial #" << m_deviceInfo.serialNumberRaw()
+                             << "(Report ID" << m_reportId << ")";
+        return; // Same data sent last time
+    }
+    m_lastSentOutputreport.clear();
+    m_lastSentOutputreport.append(data);
+
+    // Prepend the Report ID to the beginning of data[] per the API..
+    data.prepend(m_reportId);
```

Comment:
It would be more efficient to build a new array starting with the report id and then std::move() the given array into m_lastSentOutputReport.